### PR TITLE
enhancement(file sink): Automatically create missing directories

### LINKED ISSUE #1043: Automatic directory creation with file sink
It would be very useful if the file sink supported automatic directory creation.
Then it could be possible to use configuration like:
`path = "/data/logs/{{ application_name }}/{{environment}}/{{logtype}}-%Y-%m-%d.log"`
without the need to create all the directories beforehand (which in case of dynamic environments is not even possible).

### PR DESCRIPTION
**Pull request by LucioFranco**
Closes #1043

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

File: src/sinks/file/file.rs [hunk 1]
```diff
@@ -2,37 +2,48 @@ use super::Encoding;
...
 }
 
 impl File {
     pub fn new(path: Bytes, encoding: Encoding) -> Self {
         let path = BytesPath(path);
+        let parent = path.as_ref().parent().map(|p| p.to_path_buf());
 
-        let fut = file::OpenOptions::new()
+        let dir_fut = if let Some(parent) = parent {
+            future::Either::A(fs::create_dir_all(parent))
```
Comment (lukesteensen) [inline src/sinks/file/file.rs]: Will this do the create even when the directory already exists? Probably not a problem, just want to be clear.

Comment (lukesteensen) [reply src/sinks/file/file.rs]: What was the resolution here?

Comment (LucioFranco) [reply src/sinks/file/file.rs]: Oops sorry I misread this yesterday, I believe it acts like `mkdir -p` which will not return an error if the directory already exists.

Each test calls this when the directory exists and when it doesn't so its safe to run if the dir exists or not.

**Comment by lukesteensen** [comment]
[Review create missing directories](https://trello.com/c/VPHm5DKD/3-review-create-missing-directories)

File: scripts/generate/templates/docs/usage/configuration/sinks/file.md.erb [hunk 2]
```diff
@@ -14,10 +14,16 @@
 
 <%= component_sections(component) %>
 
+### Dynamic file and directory creation
+
+Vector will attempt to create the entire directory structure and the file when
+emitting events to the file sink. This requires that the Vector agent must have
+the correct permissions to create and write to files in the specified directories.
```
Comment (lukesteensen) [inline scripts/generate/templates/docs/usage/configuration/sinks/file.md.erb]: `This requires that the Vector agent must have` should probably be either `The Vector agent must have` or `This requires that the Vector agent have`

**Pull request merged**